src/System.Memory/src/System/Span.cs

```diff
@@ -0,0 +1,413 @@
+        /// </summary>
+        public static implicit operator Span<T>(ArraySegment<T> arraySegment) => new Span<T>(arraySegment.Array, arraySegment.Offset, arraySegment.Count);
+
+        /// <summary>
+        /// Forms a slice out of the given span, beginning at 'start'.
+        /// </summary>
+        /// <param name="start">The index at which to begin this slice.</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> index is not in range (&lt;0 or &gt;&eq;Length).
+        /// </exception>
+        public Span<T> Slice(int start)
+        {
+            if ((uint)start > (uint)Length)
+                throw new ArgumentOutOfRangeException(nameof(start));
+
+            IntPtr newOffset = _byteOffset.Add<T>(start);
+            int length = Length - start;
+            return new Span<T>(_object, newOffset, length);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given span, beginning at 'start', of given length
+        /// </summary>
+        /// <param name="start">The index at which to begin this slice.</param>
+        /// <param name="end">The index at which to end this slice (exclusive).</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> or end index is not in range (&lt;0 or &gt;&eq;Length).
+        /// </exception>
+        public Span<T> Slice(int start, int length)
+        {
+            if ((uint)start > (uint)Length || (uint)length > (uint)(Length - start))
+                throw new ArgumentOutOfRangeException(nameof(start));
+
+            IntPtr newOffset = _byteOffset.Add<T>(start);
+            return new Span<T>(_object, newOffset, length);
+        }
+
+        /// <summary>
+        /// Copies the contents of this span into a new array.  This heap
+        /// allocates, so should generally be avoided, however it is sometimes
+        /// necessary to bridge the gap with APIs written in terms of arrays.
+        /// </summary>
+        public T[] ToArray()
+        {
+            if (Length == 0)
+                return SpanHelpers.EmptyArray<T>();
+
+            T[] result = new T[Length];
+            CopyTo(result);
+            return result;
+        }
+
+        /// <summary>
+        /// Returns a 0-length span whose base is the null pointer.
+        /// </summary>
+        public static readonly Span<T> Empty = default(Span<T>);
+
+        /// <summary>
+        /// Returns a reference to the 0th element of the Span. If the Span is empty, returns a reference to the location where the 0th element
+        /// would have been stored. Such a reference can be used for pinning but must never be dereferenced.
+        /// </summary>
+        public ref T DangerousGetPinnableReference
+        {
+            get
+            {
+                switch (_object)
```

Comment:
This is super slow. Instead, this should be something like:

```
public struct Span<T> 
{
    [StructLayout(LayoutKind.Sequential)] 
    private sealed class Pinnable
    {
        public T Data;
    }

    private readonly Pinnable _pinnable;
    private readonly IntPtr _byteOffset;


    public ref T DangerousGetPinnableReference
    {
        get
        {
            return (_pinnable != null) ? Unsafe.AddByteOffset<T>(ref _pinnable.Data, _byteOffset) : Unsafe.AsRef<T>((void*)_byteOffset);
        }
    }
}
```

Also, I would manually inline the line into the indexer so that the JIT has less inlining to do for it.